cpp/src/arrow/filesystem/s3_internal.h

```diff
@@ -178,6 +179,17 @@ inline TimePoint FromAwsDatetime(const Aws::Utils::DateTime& dt) {
   return std::chrono::time_point_cast<std::chrono::nanoseconds>(dt.UnderlyingTimestamp());
 }
 
+template <typename ErrorType>
+S3RetryStrategy::AWSErrorDetail ErrorToDetail(
+    const Aws::Client::AWSError<ErrorType>& error) {
```

Comment:
Since this is useful only in `s3fs.cc`, not in the tests, I would define this directly inside `WrappedRetryStrategy`.